### ISSUE TITLE
[Question] How is the ExactMarginalLogLikelihood actually computed?

### ISSUE THREAD

**Issue by jarandaf** [cornellius-gp/gpytorch#683]
I am trying to use gpytorch to connect a neural network to a regular GP. The neural network learns a distance between input pairs and this distance is used to evaluate a valid kernel `K`. For that purpose, I created a custom kernel implementing the `forward` method (i.e. creating the gram matrix `K`).

I am using exact inference. During training the negative marginal likelihood loss decreases (as desired), which makes me think this is somehow working. Here comes my question: how is the `ExactMarginalLogLikelihood` actually computed? In regular GPs, the cholesky decomposition of `K` is usually computed to get the loss term. If I try to compute it after training I get non-positive semidefinite errors (i.e. the learned `K` is somehow not positive definite) but no error is actually reported during training.

Thank you in advance.

**Comment by gpleiss** [comment]
Can you include the code that you’re using to compute the MLL after training?
The procedure for computing the MLL is described in this paper: https://arxiv.org/abs/1809.11165

**Comment by jarandaf** [comment]
I use the standard MLL loss (`K` is the "learned" covariance matrix, `Xs` the train points and `ys` the target variable):

```python
K[np.diag_indices_from(K)] += (1e-8)
L = np.linalg.cholesky(K)  # this fails due to non positive definite matrix
log_det = 2.*np.sum(np.log(np.diag(L)))
P, _ = lapack.dpotrs(L, ys, lower=True)

# mll loss
-0.5*ys.T.dot(P) - 0.5*log_det - 0.5*Xs.shape[0]*np.log(2*np.pi)
```

Other libraries like `GPy` also compute it this way.

**Comment by gpleiss** [comment]
This will disagree with our MLL a bit, because we use a stochastic (unbiased) estimate of the MLL, which is much more scalable than Cholesky-based methods.

How much disagreement are you seeing?

**Comment by gpleiss** [comment]
Also, what version of GPyTorch are you using?

**Comment by jarandaf** [comment]
> This will disagree with our MLL a bit, because we use a stochastic (unbiased) estimate of the MLL, which is much more scalable than Cholesky-based methods.
> 
> How much disagreement are you seeing?

Well, the main issue here is that the learned `K` does not seem to be a valid kernel. I wanted to check this by computing its cholesky decomposition but it seems `K` is not PSD (some negative eigenvalues do pop up). At prediction time (using GPyTorch), though, no error is raised (neither during training as mentioned) and I get better results than simpler kernels (e.g. RBF) if I use some metric like RMSE in a regression setting.

The problem may be on my side with the custom kernel. Some checks could prove helpful to avoid these kind of issues, nevertheless.

> Also, what version of GPyTorch are you using?

`0.3.2`.

**Comment by jacobrgardner** [comment]
How negative are the eigenvalues you are seeing? If they are larger than -1e-4 or so, the matrix will still end up being positive definite after adding the s^2*I term to the matrix. Or in general if they aren't too negative, CG tends to still work in terms of giving solves even when Cholesky hard fails.
